### PR TITLE
feat: Map GitLab Runner Docker Parameters

### PR DESCRIPTION
**Pull request by fabitosh**
We are currently only exposing a minimal set of adjustable parameters to the REST API of runnrs. In an effort to prepare for advanced use-cases this PR maps all available Docker Executor parameters of the GitLab documentation to the glrcfg. 

[GitLab Runners Docker Section](https://archives.docs.gitlab.com/15.11/runner/configuration/advanced-configuration.html#the-runnersdocker-section)

We implement all types according to the specifications in the documentation. For parameters whose default is explicitly specified, it is taken over.

### CODE REVIEW

File: glrcfg/src/runner/executors/docker.rs [hunk 1]
```diff
@@ -10,35 +9,143 @@ macro_rules! stringvec {
...
+    pub network_mode: Option<String>,
+    pub network_mtu: u32, // written in cli runner creation, not in gitlab runner docs
+    pub mac_address: Option<String>,
+    pub oom_kill_disable: bool, // written in cli runner creation
+    pub oom_score_adjust: Option<i32>,
+    pub privileged: bool, // written in cli runner creation
+    pub pull_policy: Vec<String>,
+    pub runtime: Option<String>,
+    pub isolation: Option<String>,
+    pub security_opt: Vec<String>, // todo: serialization needs to use a : instead of the , between elements
```
Comment (fabitosh) [inline glrcfg/src/runner/executors/docker.rs:57]: Implement a serde serialization for this

File: glrcfg/src/runner/executors/docker.rs [hunk 2]
```diff
@@ -10,35 +9,143 @@ macro_rules! stringvec {
...
+    pub allowed_privileged_services: Vec<String>,
+    pub cache_dir: Option<String>,
+    pub cap_add: Vec<String>,
+    pub cap_drop: Vec<String>,
+    pub cpuset_cpus: Option<String>,
+    pub cpuset_mems: Option<String>,
+    pub cpu_shares: u32,
+    pub cpus: Option<String>,
+    pub devices: Vec<String>,
+    pub device_cgroup_rules: Vec<String>, // https://docs.docker.com/compose/compose-file/05-services/#device_cgroup_rules
```
Comment (fabitosh) [inline glrcfg/src/runner/executors/docker.rs:29]: Unsure whether that would work like this.

File: glrcfg/src/runner/executors/docker.rs [hunk 3]
```diff
@@ -10,35 +9,143 @@ macro_rules! stringvec {
...
+    pub memory: Option<String>,
+    pub memory_swap: Option<String>,
+    pub memory_reservation: Option<String>,
+    pub network_mode: Option<String>,
+    pub network_mtu: u32, // written in cli runner creation, not in gitlab runner docs
+    pub mac_address: Option<String>,
+    pub oom_kill_disable: bool, // written in cli runner creation
+    pub oom_score_adjust: Option<i32>,
+    pub privileged: bool, // written in cli runner creation
+    pub pull_policy: Vec<String>,
```
Comment (fabitosh) [inline glrcfg/src/runner/executors/docker.rs:54]: Input can actually be String as well as a list of strings.

The list allows for a chaining of multiple policies if the first one fails

File: glrcfg/src/runner/executors/docker.rs [hunk 4]
```diff
@@ -10,35 +9,282 @@ macro_rules! stringvec {
...
+        let serialized = serde_json::to_string(&policy).unwrap();
+        assert_eq!(serialized, "\"single_policy\"");
+    }
+
+    #[test]
+    fn test_pull_policy_single_none() {
+        let policy = PullPolicy::Single(None);
+        let serialized = serde_json::to_string(&policy).unwrap();
+        assert_eq!(serialized, "null");
+    }
```
Comment (cdbrkfxrpt) [inline glrcfg/src/runner/executors/docker.rs:275]:
```suggestion
```

no longer needed with the introduced changes; that aside, we wouldn't _want_ a `None` pull policy to serialize to a `null` value in JSON, would we?

File: glrcfg/src/runner/executors/docker.rs [hunk 5]
```diff
@@ -10,35 +9,282 @@ macro_rules! stringvec {
...
+            value: "garbl".to_string(),
+        };
+
+        let serialized = serde_json::to_string(&opt).unwrap();
+        assert_eq!(serialized, "\"warbl:garbl\"");
+    }
+
+    #[test]
+    fn test_pull_policy_single_some() {
+        let policy = PullPolicy::Single(Some("single_policy".to_string()));
```
Comment (cdbrkfxrpt) [inline glrcfg/src/runner/executors/docker.rs:265]:
```suggestion
        let policy = PullPolicy::Single("single_policy".to_string());
```

**Review by cdbrkfxrpt — CHANGES_REQUESTED** [review]

**Review by cdbrkfxrpt — APPROVED** [review]

**Review by cdbrkfxrpt — APPROVED** [review]

**Pull request merged**